mautic/mautic
# Segment Unit Tests

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | 
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | Y
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR adds a few much needed functional tests to the segments functionality of Mautic.

While writing the tests, I came across a bug that prevented contacts from being removed from segments when they no longer met criteria. @escopecz pointed me to a PR by @kermorgant that fixed this issue (https://github.com/mautic/mautic/pull/4110). I've merged that here and added tests that ensure it's functionality. 

These tests cover:

- [x] CLI Commands to rebuild segments
- [x] Only global segments show up in the user preference pane
- [x] Segment membership counts are correct
- [x] Segment filters adhere appropriately to `AND` & `OR` selections when building

Still to do, but by no means preventing this from being merged, is fleshing out the filter tests to test more possible combinations of filters, along with segments that could have many different filters.

#### Developer Goodies

This PR also adds the ability for developers to test against a live database using data fixtures. A data fixture allows you to inject data into the test database that will be isolated and reset between tests. To make use of data fixtures, add a `DataFixtures\ORM` directory in your bundle's `Tests` directory. You can see an example here: https://github.com/mautic/mautic/pull/4129/files#diff-9dd9eb4117bbcd2364fbb66fe71ba625

When writing your test, simply extend the `Mautic\CoreBundle\Test\MauticWebTestCase` class, and then you can access your data fixtures that you created references to by using `$this->fixtures->getReference('reference-id'). You can see an example of that here: https://github.com/mautic/mautic/pull/4129/files#diff-e437f51b2e7553085b1ec4a266b02ec0R13

The referenced object that is returned acts just like the entity you created.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Create a segment with the filter of `title` = `Mrs.`
2. Have contacts (or create them) that have the title of `Mrs.`
3. Rebuild the segment using `app/console m:s:u -i #`, changing # with the id of your created segment.
4. Verify the appropriate contacts were added.
5. Remove the `Mrs.` title from your contacts.
6. Rebuild the segments again, and see that none are removed when they should be.

#### Steps to test this PR:
1. Apply the PR
2. Go through the reproduction steps above
3. In step 6, you'll see that the contacts are removed
4. Run the unit tests

#### List backwards compatibility breaks:
1. In 2.4, we deprecated the core `MessageQueue` entity, and moved it to the ChannelBundle. This PR removes it from core a bit ahead of schedule. This was required in order to facilitate appropriate testing against a live database that was able to be isolated and refreshed between tests. It's not expected that removing this will cause issues. If you have created a plugin that integrates with the MessageQueue functionality of Mautic Marketing Messages, you may be affected if you're still using the MessageQueue entity directly.
